### DISCUSSION TITLE
Can Flare have an entity / character sprite be scaled up / be rendered 4 times as large?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#882] [Q&A] [answered]
Can Flare have an entity / character sprite be scaled up / be rendered 4 times as large and or can a character take up ~'4 tiles'? Kind of like how Baldur's Gate / Runescape / WoW have big sized enemies that take up more floor space tiles / bigger sprite circle / bigger shadow (Azuregos https://imgur.com/DLSarif + https://www.reddit.com/r/wow/comments/tbead/i_felt_bad_for_saying_stuff_like_that_to_azuregos/ ). From: https://github.com/flareteam/flare-game/issues/880#issuecomment-1012617893

**Comment by dorkster** [comment]
Not at runtime, but there's nothing stopping you from making sprites that big.

The problem is that the collisions and pathfinding are currently restricted to being one tile per entity. This has been an [issue](https://github.com/flareteam/flare-engine/issues/128) for a long while, but it just hasn't been a priority to tackle since all of flare-game's enemies are 1x1 sized.

**Reply by WithinAmnesia** [reply]
Is it possible to make Flare hit boxes other sizes than 1x1; for instance a ~2x2 / ~4 square circle? What would that take / how complex of an issue is that? How would that affect pathfinding? Also in regular Flare a ~2x2 / ~4 square circle large scale enemy would be really cool, like a dragon, a giant etc. bigger enemies could help more than just my endeavour but could make Flare have pretty cool big enemies.

**Reply by dorkster** [reply]
Hit boxes in Flare are actually backwards to what you might think. The *attacks/spells* (hazards) have a hit circle (defined by the `radius` attribute). Then, for each hazard, we iterate over the potential targets looking to see if one is in this circle. This should be easy to scale up in theory, since we would add the radius of the target onto the hazard's before doing the check.

Pathfinding is much more complicated. The A* algorithm that we use assumes that entities are 1x1 tile. I have no idea how viable it would be to make multi-tile entities work with the existing algorithm. Perhaps it would be necessary to change algorithms...

**Reply by WithinAmnesia** [reply]
Okay so in theory:
- The visuals of entry sprites can be scaled 4 times larger and work?
- The entity hit boxes / radius of target / entity can be scaled up 4 times and work?
- The pathfinding "A* algorithm" assumes / works / is built with 1x1 tile size entities, it is currently unknown / untested to see how it works with larger / 2x2 tile sized / scale entities.

"A* algorithm" 2x2 tile sized entity pathfinding: So a rough / crash course test could be performed to gauge / see how the current "A* algorithm" works in handling 2x2 tile entities? For that potential 2x2 tile entity "A* algorithm" performance data it could lead to how / where / what to tweak / change / add to "A* algorithm" to work if needed?

If needed could the A* algorithm be improved / added onto to work with 2x2 tile entities if "A* algorithm" currently is not up to the task? Hopefully without too much added complexity, if 2x2 tile entity "A* algorithm" path finding can somehow work; could both 1x1 and 2x2 pathfinding work nice together or is this an all or nothing with "A* algorithm" working on either 1x1 or 2x2?

Could multiple sizes on tile entities work hopefully in the future (without too many problems) such as 1x1, 2x2, 3x3 etc. ? Or if 2x2 tile entity pathfinding with "A* algorithm" works, will all entities need to be 2x2 as well; or is this an added feature where 1x1 and 2x2 could work together?

If 2x2 tile entity size pathfinding works; would that be the last step in order to effectively get 'Flare x 4' working? Or is there another issue I am not considering? Flare x 4 (128x128) icons and inventory works, Flare x 4 menus effectively work and should fully work in theory, Flare x 4 (4x4 tile entities) sprite rendering works in theory, Flare x 4 tileset should work in theory with tiles effectively being split in 4 and still working with x 4 enlarged tile asset art, Flare x 4 spell effect animations / particle effects should work in theory (maybe)? Can a x 4 times larger Fire ball animation art still work? The hazard radius works so if the x 4 enlarged art should also hopefully work?

I have so many points of intrigue that I do not intrinsically understand. Yet for all that I do not innately comprehend herein; such points of interest all the same fascinate my curiosity to better understand how the Flare engine works and thinks XD. Thank you for your time and explanations @dorkster :-D.

**Reply by dorkster** [reply]
If you want to scale up the entities (player/npcs/enemies), you're going to want to scale up all the map tiles as well. Otherwise, all the entities will be giants next to the existing map tiles. If you scale everything up, the relative sizes stay the same. A tile is a tile. You would just be making the tiles bigger. If you look at `mods/fantasycore/engine/tileset_config.txt`, you'll see that the tile size we use is 64x32. So 4x of that would be 256x128. That would be what you are targeting.

In short, the no changes to pathfinding would be necessary and is really unrelated to your original question. You also wouldn't need to change the hazards' hit radius, which use tiles as their unit of measurement.

**Reply by WithinAmnesia** [reply]
Okay so with 256x128 scale tiles, Pathfinding would not need to be changed, the entities can be scaled up x 4 and work, the radious would not need to be changed as well?

If this is true I should start by making 256x128 (x 4 enlarged tilesets) and make a 256 x 128 scale Flare map in Tiled.exe and test out the bigger tiles?

Also how do I x 4 enlarge the entities and spell / ability animations?

**Reply by WithinAmnesia** [reply]
Okay I got a test 256 x 128 Tileset / 1000 x 1000 Map working but I am not sure how to scale up the character entity by x 4 to match (nor how to x 4 enlarge the npc and enemy entities and spell / ability animations). Is there a process of going to the character / entity art and enlarging by x 4 and doing the same process of implementing x 4 enlarged art assets with the Flare engine as with the icons and tileset / 256 x 128 tile map to match the same scale overall for all art assets / components?
https://cdn.discordapp.com/attachments/872743295003000852/931828402300780605/256x128_Tilesets_Work.png
Also I wonder why is the minimap is gone?

**Reply by dorkster** [reply]
Before moving on to entities, you should make sure you update the tileset definition if you haven't already. For the dungeon set, this is in `mods/fantasycore/tilesetdefs/tileset_dungeon.txt`. Should just be a matter of multiplying that relevant values by 4. Here's what parts of the the `tile` and `animation` properties that you'd need to change:

```
tile=265,800,0,128,64,32,32
         ^^^^^^^^^^^^^^^^^^
animation=265;800,0,66ms;334,128,66ms;462,128,66ms;590,128,66ms;
              ^^^^^      ^^^^^^^      ^^^^^^^      ^^^^^^^
```

The process should be similar for the character/entity art, except instead of a tileset definition you will edit the animation files. Most of the animation definitions are in `mods/fantasycore/animations`, but there are also some in `mods/empyrean_campaign/animations` as well. Again, here's an example of the values you'd need to multiply by 4:

```
frame=0,0,68,249,27,33,8,39
          ^^^^^^^^^^^^^^^^^
```

>Also I wonder why is the minimap is gone?

Perhaps you hit the 'M' key by accident? There was actually bug with the dev console where this shortcut wasn't blocked, so the minimap would change modes whenever the M key was typed into the command box. You can also edit the minimap setting on the Interface page of the config menu. Possible values are: Hidden, Visible, and Visible (2x). I may make it to that the M key only cycles the visible modes to prevent people from accidentally hiding it.

**Reply by WithinAmnesia** [reply]
Thank you for sharing, the tilesetdef was changed like above to get the map to not crash, small tile art with small tileset_defs on big tiles ~'works' / has lots of black in-between the tiles and does not crash (but looks super buggy); although big tile art with small tileset_defs on big tiles do not even run and just crash / boot to game start area XD. I guessed what numbers need to be changed on the tileset def although now I know / better understand what should be changed to scale up. I did not test tileset / map art animations and I am glad to learn what should be done with scaling up tileset / map animations :-).

I will start enlarging the default avatar animations; lots of lines :-P.

The minimap is back on a new start up, it looks like yeah the dev console where exec intermap=maps/256x128_test0+.txt ; two "M" key presses so it will cycle and potentially with any more "M" presses throw the minimap all over the place XD.

**Comment by WithinAmnesia** [comment, marked as answer]
Yes. Yet only on a 64-bit executable; I hope: https://github.com/flareteam/flare-game/issues/885#issuecomment-1014059725 . Yet once a Windows 64-bit executable is available there will need to be a stress test to see if it will work with also enlarged x 4 menus, maps, entities and other art assets.
